cinema/serializers.py

```diff
@@ -68,13 +78,58 @@ class Meta:
             "movie_title",
             "cinema_hall_name",
             "cinema_hall_capacity",
+            "tickets_available"
         )
 
 
 class MovieSessionDetailSerializer(MovieSessionSerializer):
     movie = MovieListSerializer(many=False, read_only=True)
     cinema_hall = CinemaHallSerializer(many=False, read_only=True)
+    taken_places = serializers.SerializerMethodField()
 
     class Meta:
         model = MovieSession
-        fields = ("id", "show_time", "movie", "cinema_hall")
+        fields = ("id", "show_time", "movie", "cinema_hall", "taken_places",)
+
+    def get_taken_places(self, obj):
+        tickets = Ticket.objects.filter(movie_session=obj)
+        return [{"row": ticket.row, "seat": ticket.seat} for ticket in tickets]
+
+
+class TicketSerializer(serializers.ModelSerializer):
+
+    class Meta:
+        model = Ticket
+        fields = ("id", "row", "seat", "movie_session")
+
+    def validate(self, attrs):
+        Ticket.validate_row_and_seat(
+            attrs["row"],
+            attrs["movie_session"].cinema_hall.rows,
+            attrs["seat"],
+            attrs["movie_session"].cinema_hall.seats_in_row,
```

Comment:
In the `validate` method of `TicketSerializer`, ensure that `attrs` contains the `movie_session` key before accessing it. Otherwise, this could lead to a `KeyError` if `movie_session` is not provided in the input data. Consider adding a check or using `attrs.get('movie_session')` to safely access it.